### PR TITLE
Bug split optimization

### PR DESCRIPTION
**Pull request by scxiao**
This issue surfaced for the resize linear mode

### CODE REVIEW

File: src/simplify_algebra.cpp [hunk 1]
```diff
@@ -421,9 +439,18 @@ struct find_splits
                 if(it == split->outputs().end())
                     break;
                 assert((*it)->name() != "slice");
-                // If there is a duplicate bail
```
Comment (pfultz2) [inline src/simplify_algebra.cpp:424]: Why is this comment removed?

File: src/simplify_algebra.cpp [hunk 2]
```diff
@@ -421,9 +439,18 @@ struct find_splits
...
+
                 if(contains(group, *it))
                     return {};
+
+                // there are should be no dependent between instructions in the group
+                if(std::any_of(group.begin(), group.end(), [&](auto i) {
+                       return is_dependent(*it, i) or is_dependent(i, *it);
```
Comment (pfultz2) [inline src/simplify_algebra.cpp:448]: Can't the `contains(group, *it)` be removed?

File: src/simplify_algebra.cpp [hunk 3]
```diff
@@ -421,9 +439,18 @@ struct find_splits
                 if(it == split->outputs().end())
                     break;
                 assert((*it)->name() != "slice");
-                // If there is a duplicate bail
+
                 if(contains(group, *it))
                     return {};
+
+                // there are should be no dependent between instructions in the group
```
Comment (pfultz2) [inline src/simplify_algebra.cpp:446]: dependent -> dependency

File: src/simplify_algebra.cpp [hunk 4]
```diff
@@ -403,6 +404,23 @@ struct find_splits
...
+        return fix<bool>([&](auto self, auto ins) -> bool {
+            if(ins == ins2)
+                return true;
+
+            if(contains(traversed, ins))
+                return false;
+
+            traversed.insert(ins);
+            const auto& inputs = ins->inputs();
+            return std::any_of(inputs.begin(), inputs.end(), [&](auto in) { return self(in); });
```
Comment (pfultz2) [inline src/simplify_algebra.cpp:420]: Wont this traverse into the parent module?

File: src/simplify_algebra.cpp (hunk 1 above)
Comment (scxiao) [reply src/simplify_algebra.cpp:424]: put it back

File: src/simplify_algebra.cpp (hunk 2 above)
Comment (scxiao) [reply src/simplify_algebra.cpp]: removed

File: src/simplify_algebra.cpp (hunk 3 above)
Comment (scxiao) [reply src/simplify_algebra.cpp]: changed

File: src/simplify_algebra.cpp (hunk 4 above)
Comment (scxiao) [reply src/simplify_algebra.cpp]: add the check whether input is in the same module

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**